src/base/utils/tar.cpp

```diff
@@ -0,0 +1,298 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2020  sledgehammer999 <hammered999@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "tar.h"
+
+#include <QByteArray>
+#include <QCoreApplication>
+#include <QHash>
+#include <QString>
+
+#include "base/logger.h"
+#include "base/utils/bytearray.h"
+#include "base/utils/fs.h"
+
+namespace
+{
+    const int TAR_HEADER_SIZE = 512;
+
+    const int TAR_FILENAME_OFFSET = 0;
+    const int TAR_FILESIZE_OFFSET = 124;
+    const int TAR_CHECKSUM_OFFSET = 148;
+    const int TAR_FILETYPE_OFFSET = 156;
+    const int TAR_FILENAME_PREFIX_OFFSET = 345;
+
+    const int TAR_FILENAME_SIZE = 100;
+    const int TAR_FILESIZE_SIZE = 12;
+    const int TAR_CHECKSUM_SIZE = 8;
+    const int TAR_FILENAME_PREFIX_SIZE = 155;
+
+    class TarEntry
```

Comment:
This class has inconsistent internal behavior. Why do you parse header partially when create instance and then parse the rest on demand? I see no point in this.

Reply:
>Why do you parse header partially when create instance and then parse the rest on demand?

What do I do partially? You mean the chechsum check and null check?

If I parse the rest on ctor, that means I would have to have extra member variables to hold a copy of the data.

Reply:
>If I parse the rest on ctor, that means I would have to have extra member variables to hold a copy of the data.

What's the problem with it? If header is invalid (or null) they will be empty. If header is valid you still copy this data when access it from caller side (since it's pointless to create this object and then never access its fields).

Reply:
>What's the problem with it?

Increased class size.
I want to stay with the parse-on-demand approach and not have unnecessary copies of data laying around. Something akin to `bedecode_token/bedecode_node` (somewhat inspired by [this](https://blog.libtorrent.org/2015/03/bdecode-parsers/))

Usually the user of the class will call only once each method and store the result. So why make the class also store a copy?

Reply:
>Usually the user of the class will call only once each method and store the result.

No problem (unless you need to call some of them twice).
But in this case the class itself becomes redundant (now it is like a "thing in itself" that only takes care of its own invariant in each method). You can easily apply any parsing function on header data separately.